streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1096,7 +1096,7 @@ private Optional<String> removeStreamThread(final long timeoutMs) throws Timeout
                 // make a copy of threads to avoid holding lock
                 for (final StreamThread streamThread : new ArrayList<>(threads)) {
                     final boolean callingThreadIsNotCurrentStreamThread = !streamThread.getName().equals(Thread.currentThread().getName());
-                    if (streamThread.isAlive() && (callingThreadIsNotCurrentStreamThread || getNumLiveStreamThreads() == 1)) {
+                    if (callingThreadIsNotCurrentStreamThread || getNumLiveStreamThreads() == 1) {
```

Comment:
@wcarlson5 Could you please help me understanding the implications of removing this check? 
My take is that: 

Without the check:
- If the stream thread is not alive and its state is not `DEAD`, it will not be removed from the list of threads. 
- If the stream thread is not alive and its state is `DEAD`, it will be removed from the list of threads.
- If the stream thread is alive and its state is not `DEAD`, it will be removed from the list of threads once it reaches `DEAD` -- here, in `getNumLiveStreamThreads()`, or in `getNextThreadIndex()`. 
- If the stream thread is alive and its state is `DEAD`, it will be removed from the list of threads.

With the check:
- If the stream thread is not alive and its state is not `DEAD`, it will not be removed from the list of threads. 
- If the stream thread is not alive and its state is `DEAD`, it will be removed from the list of threads in `getNumLiveStreamThreads()` or in `getNextThreadIndex()` but not here.
- If the stream thread is alive and its state is not `DEAD`, it will be removed from the list of threads once it reaches `DEAD` -- here, in `getNumLiveStreamThreads()`, or in `getNextThreadIndex()`. 
- If the stream thread is alive and its state is `DEAD`, it will be removed from the list of threads.

So it seems the check does not change much except that it delays the removal from the list if the stream thread is not alive and its state is `DEAD`.

Did I miss a case?

For context, this check blocks the migration to Mockito since Mockito cannot stub native methods and `isAlive()` is a native method.

Reply:
So this check is important and I don't think we can just remove it.

The important thing this is doing is preventing a thread from being removed that is already dead. If a thread is removed but not finished cleaning up it is still in the thread list. If we then call remove threads and that dead thread is picked then no thread will really be removed so that call is broken. Even worse though the cache size of all the other threads are changed to some value that is not correct.

Instead of just removing the check you can maybe change to do something like 
`thread.state() == StreamThread.State.DEAD`

Maybe you can add a helper method to check that instead of using the native methods if w really need to.

On a side note it is concerning to me that we can't mock native methods anymore that does not seem a good thing. Why are we removing this ablity?

Reply:
@wcarlson5 Thanks for the info! Really useful! 

> Instead of just removing the check you can maybe change to do something like
> thread.state() == StreamThread.State.DEAD

I guess you mean `thread.state() != StreamThread.State.DEAD && thread.state() != StreamThread.State.PENDING_SHUTDOWN`
I think we also need to add `thread.state() != StreamThread.State.PENDING_SHUTDOWN` since stream threads in `StreamThread.State.PENDING_SHUTDOWN` are skipped in `getNumLiveStreamThreads()` and the result of that method is used to resize the cache. 

I am not too much concerned about losing the ability to mock native methods. I think it is much more important to use a mocking framework that is actively maintained and does not block Java version upgrades in AK. As far as I understand that is the reason we move away from EasyMock and PowerMock to Mockito.

Reply:
You can look at the is alive method for state checks that are happening and refactor it into a method that works for Mockito. 

Ah that makes more sense. Thanks for clarifying

Reply:
Wouldn't it be possible to just call `streamThread.state().isAlive()`?

Reply:
Maybe, if the mock works then proabably

Reply:
That is not a native method. A native method looks like the following in `java.lang.Thread`:
```
public final native boolean isAlive();
```

Reply:
Ah, I think you mean because it is a method on an enum.

Reply:
Thank you both for the discussion here.

I believe that we cannot use `streamThread.state().isAlive()` here because a thread in `CREATED` state is not considered `isAlive()` by this method which can lead to the following regression from existing behaviour:
1. Inside `addStreamThread()`, thread gets created and added to `threads` list at `createAndAddStreamThread()` but Thread.start() (which transitions the thread state to STARTING) hasn't been called yet (because, let's say current thread is waiting at line 1033 in addStreamThread() method).
2. `removeStreamThread` gets called by another thread.

**Old behaviour:**

`streamThread.isAlive()` will return true and the thread will be removed from the list. When the first thread 

**New behaviour:**

`streamThread.state().isAlive()` will return false and the thread will not be removed from the list, hence `removeStreamThread` will not remove this thread.

I believe that this new behaviour is a change in public interface contract for `removeStreamThread` and hence, perhaps could be a breaking change.

Alternatively, 
the simplest solution is to create a new method which is a wrapper over the native method such as:
```
// Ensure Mockito can stub method for KafkaStreamTest.
    public boolean isStateAlive() {
        return state.isAlive();
    }
```
Is that something that would be acceptable to you folks?

Reply:
Since we are talking about the migration from EasyMock to Mockito and not about improving the current behavior, I would go with the simplest solution. 

I think, you meant to write:
```java
public boolean isThreadAlive() {
    return isAlive();
}
```
 
Thank you for proposing that solution!